cachito/workers/pkg_managers/general_js.py

```diff
@@ -457,3 +460,130 @@ def upload_non_registry_dependency(
+
+    # The name of the dependency
+    # In package-lock.json these are the keys in the "dependencies" object
+    # In yarn.lock, these are the top-level keys in the file (the keys are <name>@<version>)
+    name: str
+
+    # The source of the dependency, i.e. resolved url or relative path.
+    # In package-lock.json, this is either the "resolved" key or the "version" key
+    # In yarn.lock, this is either the "resolved" key or the filepath in the top-level key
+    source: str
+
+    # The actual semver version of the dependency
+    # In package-lock.json, this either the "version" key or not present
+    # In yarn.lock, this is always the "version" key
+    version: str = None
+
+    # The hash algorithm and base64-encoded checksum of the dependency
+    # In package-lock.json, this is the "integrity" key and will always be present for registry
+    # deps and tarball deps (it is not relevant for the other types).
+    # In yarn.lock, the "integrity" key seems to be present only for registry deps by default, but
+    # all resolved urls also have a SHA1 checksum in the fragment part. It is unclear how yarn
+    # decides whether to check the integrity value or the checksum in the url when both are present.
+    integrity: str = None
+
+
+def process_non_registry_dependency(js_dep):
+    """
+    Convert the input dependency not from the NPM registry to a Nexus hosted dependency.
+
+    :param JSDependency js_dep: the dependency to be converted
+    :return: information about the replacement dependency in Nexus
+    :rtype: JSDependency
+    :raise CachitoError: if the dependency is from an unsupported location
+        or has an unexpected format
+    """
+    git_prefixes = {
+        "git://",
+        "git+http://",
+        "git+https://",
+        "git+ssh://",
+        "github:",
+        "bitbucket:",
+        "gitlab:",
+    }
+    http_prefixes = {"http://", "https://"}
+    verify_scripts = False
+    checksum_info = None
+    if any(js_dep.source.startswith(prefix) for prefix in git_prefixes):
+        try:
+            _, commit_hash = js_dep.source.rsplit("#", 1)
+        except ValueError:
+            msg = f"The dependency {js_dep.source} was in an unexpected format"
+            log.error(msg)
+            raise CachitoError(msg)
+        # When the dependency is uploaded to the Nexus hosted repository, it will be in the format
+        # of `<version>-gitcommit-<commit hash>`
+        version_suffix = f"-external-gitcommit-{commit_hash}"
+        # Dangerous scripts might be required to be executed by `npm pack` since this is a Git
+        # dependency. If those scripts are present, Cachito will fail the request since it will not
+        # execute those scripts when packing the dependency.
+        verify_scripts = True
+    elif any(js_dep.source.startswith(prefix) for prefix in http_prefixes):
+        if not js_dep.integrity:
+            msg = f"The dependency {js_dep.source} is missing the integrity value"
```

Comment:
Should this also reflect if the dependency is coming from package-lock.json or yarn.lock? Because yarn.lock has another possible source for integrity. I kind of wanted to avoid that with `JSDependency`.
Would something like this be good enough?
```
Is the "integrity" key missing in your lockfile?
```
(completely ignoring the fact that in yarn.lock, integrity typically comes from urls)

Reply:
Sounds good!